default.nix

```diff
@@ -125,6 +125,31 @@ let
     ${concatStrings (imap (index: create-f (q // { inherit index; })) x.partitions)}
   '';
 
+  create.zfs = q: x: ''
+    ZFSDEVICES_${x.pool}="''${ZFSDEVICES_${x.pool}:-}${q.device} "
+  '';
+
+  create.zfs_filesystem = q: x: ''
+    zfs create ${q.pool}/${x.name} \
+      ${lib.optionalString (isAttrs x.options or null) (concatStringsSep " " (mapAttrsToList (n: v: "-o ${n}=${v}") x.options))}
+  '';
+
+  create.zfs_volume = q: x: ''
+    zfs create ${q.pool}/${x.name} \
+      -V ${x.size} \
+      ${lib.optionalString (isAttrs x.options or null) (concatStringsSep " " (mapAttrsToList (n: v: "-o ${n}=${v}") x.options))}
+    udevadm trigger --subsystem-match=block; udevadm settle
+    ${create-f { device = "/dev/zvol/${q.pool}/${x.name}"; } x.content}
+  '';
+
+  create.zpool = q: x: ''
+    zpool create ${q.name} \
+      ${lib.optionalString (!isNull x.mode || x.mode != "stripe") x.mode} \
+      ${lib.optionalString (isAttrs x.options or null) (concatStringsSep " " (mapAttrsToList (n: v: "-o ${n}=${v}") x.options))} \
```

Comment:
Is there a way to set filesystem options with -O on the root dataset?

Reply:
do we need that? what would be the difference to just -o on the zpool?

Reply:
`zpool create` will create the root zfs dataset, without -O these dataset properties can only be set with `zfs set` afterwards. `zpool create -o` does only set pool properties and not dataset properties.